### DISCUSSION TITLE
Timestamp delay issue with DigitalOutput node

### DISCUSSION THREAD

**Discussion by cspurandare** [bonsai-rx/bonsai#1483] [Q&A] [answered]
Hi,
I have encountered an issue with the DigitalOutput node, which I am using to write TTLs to an Arduino (which later relays them to Open Ephys for syncing). The issue arises when the DigitalOutput node is called upon, without any delay from the start of the Bonsai script (right after the green button is pressed). The issue goes away, if there is a delay of around 5 seconds. I used a simple chain as shown in the images attached, and am able to recreate the issue when delay is zero in the first node, and resolve it with a delay of 5 seconds. The timestamp difference between Bonsai processing before and after the "DigitalOutput" node is substantial (>2 seconds), but this remains constant through the recording. My hypothesis is that this is the rough amount of time Bonsai takes to establish communication with an Arduino, and that causes the timestamps to get shifted, when I use the zip function. 
Is this a known issue and is there any other package update or other method to resolve it, other than using delays?
Thanks,
Chinmay
![Screenshot 2023-07-19 180027](https://github.com/bonsai-rx/bonsai/assets/39069923/1ee6be89-58cc-4e71-82d9-a58eaa257e47)
![Screenshot 2023-07-19 180125](https://github.com/bonsai-rx/bonsai/assets/39069923/fb248ad7-06a5-44dd-98c2-7e74c168550d)

**Comment by bruno-f-cruz** [comment, marked as answer]
Indeed this behavior is probably related with the initialization of bonsai+arduino. If you want to reduce the latency, you can first instantiate the Arduino object and use subjects to trigger the pulse ([see this tutorial](https://github.com/bruno-f-cruz/docs-wip/blob/%2320/articles/ArduinoArticle.md)).

That said, you should not use the timestamp provided by bonsai to align the the digital output with ephys. The timestamp uses the OS clock, which, as opposed to a microcontroller is not a real-time operating system.

**Comment by cspurandare** [comment]
Hi Bruno,
Thank you for your prompt answer. I have marked it as the answer, but I had a follow up question about the use of Bonsai Timestamp. Could you elaborate what you meant by "not using timestamp provided by bonsai to align". I have noticed that when the Arduino is properly initialized, there is a <2ms delay between the timestamps before and after sending the TTL to arduino. My bonsai setup requires the synchronization between a camera, another arduino reading serial data and sometimes display monitors. I was hoping to use the timestamps on Bonsai end for all of these behavioral measures, and also save the timestamps of Arduino TTL generation. Then I was hoping to use those timestamps, and the "sample_number" file in OpenEphys events folder to sync Bonsai and OpenEphys. My full workflow is attached, do you think this is kosher or am I missing something?
Thanks in advance,
Chinmay 

https://drive.google.com/file/d/1vUKrtjB7YLfVvkIFSM1lXZWYw7LcsGrV/view?usp=drive_link, https://drive.google.com/file/d/1ewIk8jjGQ3wikIkw-W6MXnFwFff9PxJX/view?usp=drive_link

**Reply by bruno-f-cruz** [reply]
To make a long story short(er), and will probably sacrifice a bit on the technical details.
The timestamps provided by bonsai, using the `Timestamp` operator are given by the operating system. Time in the operating system is somewhat "non-linear", in the sense that it might jitter in the order of a few dozens of milliseconds. On top of that, you should also consider the jitter that is incurred when timestamping events in "software". Say, your camera acquires a frame and sends it to the PC. The time it takes to send that frame, arrive to the PC, be parsed by Bonsai, and timestamped will likely also be jittered, and depend on a lot of factors (e.g. the current processing load). As a result, if you want to align physical events you should rely on the synchronization across devices that do implement some sort of real-time operating system. For instance, if you want to align the display of the monitor with open ehys, you should rely on physically measuring the output of the screen (e.g. with a photodiode) rather than using the timestamp provided by bonsai for that frame. The same is true for Arduino: if you want to align an Arduino, you should use the timestamp provided by the microcontroller and NOT the one provided by Bonsai.

Of course, under certain conditions, these software timestamps might be sufficient, but since you are using an ephys system, I assume that a few dozen milliseconds of jitter is likely a killer for your experiment...

Hope it helps,
Bruno!
